File: gateway/lua/ez.lua
local cjson = require "cjson"

local ez_client = require("ez_client")

local env = require("lua/env")
local log = require("lua/log")

-- private
log.info("setting requester")
local requester = ez_client.new_requester(
   env.EZ_INPUT_HOST, env.EZ_INPUT_PORT
)

-- public

local M = {}
function M.r(service, url, body, opts)
   local body = cjson.encode(body)
   local frames = {service, url, body}
   local res, err = requester(frames, opts)
   if err then
      log.err(tostring(err))
      return nil, {code = 500, msg = "server error"}
   end
   local ok, res_body = unpack(res)
   res_body = cjson.decode(res_body)
   if res_body == "" or res_body == ngx.null then
      res_body = nil
   end
   if ok == "ERR" then
      return nil, res_body
   end
   return res_body, nil
end

function M.r_json(service, url, body, opts)
   local res, err = M.r(service, url, body, opts)
   if err or not res then
      return res, err
   end
   local res, decode_err = cjson.decode(res)
   if decode_err then
      return nil, {code = 500, msg = "server error"}
   end
   return res, nil
end

return M

File: workers/ez.py
import functools
import json
import logging
from typing import Dict
import os

from ez_arch_worker.api import EzClient
from ez_arch_worker.api import new_client


EZ_INPUT_HOST: str = os.environ.get("EZ_INPUT_HOST", "localhost")
EZ_INPUT_PORT: int = int(os.environ.get("EZ_INPUT_PORT", 9000))


class EzRequester:
    client: EzClient

    async def r(
            self,
            service_name: bytes,
            path: bytes,
            body: Dict,
            die_on_err: bool = True
    ):
        frames = [service_name, path, json.dumps(body).encode("utf-8")]
        ok, res = await self.client.r(frames)
        if ok == b"ERR":
            logging.error("request %s::%s failed: %s",
                          service_name, path, res)
            if die_on_err:
                raise Exception(res)
            return (None, json.loads(res))
        return (json.loads(res), None)

    async def r_json(
            self,
            service_name: bytes,
            path: bytes,
            body: Dict,
            die_on_err: bool = True
    ):
        res, err = await self.r(service_name, path, body, die_on_err)
        if res and not err:
            res = json.loads(res)
        return res, err


async def new_requester() -> EzRequester:
    requester = EzRequester()
    requester.client = await new_client(EZ_INPUT_HOST, EZ_INPUT_PORT)
    return requester


def endpoint(route: bytes = b"/"):
    def wrap(_fn):
        @functools.wraps(_fn)
        def impl_fn(*args, **kwargs):
            return _fn(*args, **kwargs)
        impl_fn.route = route
        return impl_fn
    return wrap
